router/batchrouter/asyncdestinationmanager/marketo-bulk-upload/marketobulkupload.go

```diff
@@ -0,0 +1,357 @@
+	}
+	return marketoBulkUpload, nil
+}
+
+func (b *MarketoBulkUploader) Poll(pollInput common.AsyncPoll) common.PollStatusResponse {
+	payload, err := json.Marshal(pollInput)
+	if err != nil {
+		b.logger.Errorf("Error in Marshalling Poll Input: %v", err)
+		return common.PollStatusResponse{
+			StatusCode: 500,
+			HasFailed:  true,
+		}
+	}
+	bodyBytes, transformerConnectionStatus := misc.HTTPCallWithRetryWithTimeout(b.TransformUrl+b.PollUrl, payload, b.timeout)
+	if transformerConnectionStatus != 200 {
+		return common.PollStatusResponse{
+			StatusCode: transformerConnectionStatus,
+			HasFailed:  true,
+		}
+	}
+	var asyncResponse common.PollStatusResponse
+	err = json.Unmarshal(bodyBytes, &asyncResponse)
+	if err != nil {
+		// needs to be retried
+		b.logger.Errorf("Error in Unmarshalling Poll Response: %v", err)
+		return common.PollStatusResponse{
+			StatusCode: 500,
+			HasFailed:  true,
+		}
+	}
+	return asyncResponse
+}
+
+func (b *MarketoBulkUploader) GenerateFailedPayload(jobs []*jobsdb.JobT, importID, destType, csvHeaders string) []byte {
+	var failedPayloadT common.AsyncFailedPayload
+	failedPayloadT.Input = make([]map[string]interface{}, len(jobs))
+	failedPayloadT.Config = b.destinationConfig
+	for index, job := range jobs {
+		failedPayloadT.Input[index] = make(map[string]interface{})
+		var message map[string]interface{}
+		metadata := make(map[string]interface{})
+		err := json.Unmarshal([]byte(common.GetTransformedData(job.EventPayload)), &message)
+		if err != nil {
+			b.logger.Errorf("Error in Unmarshalling GetUploadStats Input : %v", err)
+			return nil
+		}
+		metadata["job_id"] = job.JobID
+		failedPayloadT.Input[index]["message"] = message
+		failedPayloadT.Input[index]["metadata"] = metadata
+	}
+	failedPayloadT.DestType = strings.ToLower(destType)
+	failedPayloadT.ImportId = importID
+	failedPayloadT.MetaData = common.MetaDataT{CSVHeaders: csvHeaders}
+	payload, err := json.Marshal(failedPayloadT)
+	if err != nil {
+		b.logger.Errorf("Error in Marshalling GetUploadStats Input : %v", err)
+		return nil
+	}
+	return payload
+}
+
+func (b *MarketoBulkUploader) GetUploadStats(UploadStatsInput common.GetUploadStatsInput) common.GetUploadStatsResponse {
+	failedJobUrl := UploadStatsInput.FailedJobURLs
+	parameters := UploadStatsInput.Parameters
+	importId := gjson.GetBytes(parameters, "importId").String()
+	csvHeaders := gjson.GetBytes(parameters, "metadata.csvHeader").String()
+	payload := b.GenerateFailedPayload(UploadStatsInput.ImportingList, importId, b.destName, csvHeaders)
+	if payload == nil {
+		return common.GetUploadStatsResponse{
+			StatusCode: 500,
+		}
+	}
+
+	failedBodyBytes, statusCode := misc.HTTPCallWithRetryWithTimeout(b.TransformUrl+failedJobUrl, payload, b.timeout)
+	if statusCode != 200 {
+		return common.GetUploadStatsResponse{
+			StatusCode: statusCode,
+		}
+	}
+	var failedJobsResponse map[string]interface{}
```

Comment:
Instead of `Unmarshal` to a `map[string]interface{}`, why not use a struct:
```go
type failedResponsePayload struct{
     status Int
     metadata struct {
         failedKeys []int
         warningKeys []int
         succeededKeys []int
}
```

```suggestion
	var failedJobsResponse failedResponsePayload
```

It will simplify error handling, avoid reflections and improve code readability

Reply:
hi, for marketo bulk upload we will raise a separate PR. Here we have just moved the [existing code](https://github.com/rudderlabs/rudder-server/blob/e20976df8243a60ca64b02ea10f436986ef6064e/router/batchrouter/handle_async.go#L137) into new structure